src/quarot/modeling_llama.py

```diff
@@ -59,17 +71,51 @@ def forward(self, x):
 
 
 class QuarotFP16LlamaFlashAttention2(LlamaFlashAttention2):
-    def __init__(self, config, act_quantization=False, online_had=False, *args, **kwargs):
+    def __init__(
+        self,
+        config,
+        act_bits: int = 16,
+        act_clip_ratio: float = 1.0,
+        k_bits: int = 16,
+        k_clip_ratio: float = 1.0,
+        k_groupsize: int | None = None,
+        v_bits: int = 16,
+        v_clip_ratio: float = 1.0,
+        v_groupsize: int | None = None,
+        online_had=False,
+        *args,
+        **kwargs,
+    ):
         super().__init__(config, *args, **kwargs)
-        self.input_quantizer = DummyActQuantizer()
-        self.act_quantization = act_quantization
         self.online_had = online_had
         self.q_proj = QuarotFP16Linear.like(self.q_proj)
         self.k_proj = QuarotFP16Linear.like(self.k_proj)
         self.v_proj = QuarotFP16Linear.like(self.v_proj)
         self.o_proj = QuarotFP16Linear.like(self.o_proj)
         self.online_o_proj_hadamard = OnlineHadamard(self.num_heads)
-        self.o_proj_input_quantizer = DummyActQuantizer()
+        self.online_k_hadamard = OnlineHadamard(self.head_dim)
+        self.online_q_hadamard = OnlineHadamard(self.head_dim)
+
+        if act_bits < 16:
+            self.input_quantizer = ActQuantizer(act_bits, symmetric=True, clip_ratio=act_clip_ratio)
+            self.o_proj_input_quantizer = ActQuantizer(act_bits, symmetric=True, clip_ratio=act_clip_ratio)
+        else:
+            self.input_quantizer = DummyActQuantizer()
```

Comment:
DummyActQuantizer is a noop with some formatting for compatibility with rest of the workflow?

Reply:
That's right!